### PR TITLE
make defining and testing webservices easier

### PR DESCRIPTION
**Pull request by bartfastiel**
* reduce complexity - keep it simple!
* using immutables
* use return values instead of mutable parameters

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/WebService.java [hunk 1]
```diff
@@ -94,9 +94,12 @@
  * provided by {@link RequestHandler#handle(Request, Response)}.
  *
  * @since 4.2
+ *
+ * @deprecated since 6.5 use {@link WebServiceDefinition}
  */
 @ServerSide
 @ExtensionPoint
+@Deprecated
 public interface WebService extends Definable<WebService.Context> {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/WebService.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/WebService.java [hunk 2]
```diff
@@ -116,6 +119,10 @@ public NewController createController(String path) {
       return new NewController(this, path);
     }
 
+    /**
+     * @deprecated since 6.5 use {@link #addController(Controller)}
+     */
+    @Deprecated
     private void register(NewController newController) {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/WebService.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java [hunk 3]
```diff
@@ -60,11 +62,22 @@
 
   private final WebService.Context context;
 
-  public WebServiceEngine(WebService[] webServices) {
+  public WebServiceEngine(WebService[] webServices, WebServiceDefinition[] webServiceDefinitions) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java]: MINOR Remove this use of "WebService"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java [hunk 4]
```diff
@@ -60,11 +62,22 @@
...
     context = new WebService.Context();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java]: MINOR Remove this use of "WebService"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java [hunk 5]
```diff
@@ -60,11 +62,22 @@
...
     for (WebService webService : webServices) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java]: MINOR Remove this use of "WebService"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java [hunk 6]
```diff
@@ -60,11 +62,22 @@
 
   private final WebService.Context context;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java]: MINOR Remove this use of "WebService"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java [hunk 7]
```diff
@@ -60,11 +62,22 @@
...
+    Arrays.stream(webServiceDefinitions)
+      .map(WebServiceDefinition::define)
+      .forEach(context::addController);
+  }
+
+  /**
+   * @deprecated since 6.5 use {@link #WebServiceEngine(WebService[],WebServiceDefinition[])}
+   */
+  @Deprecated
+  public WebServiceEngine(WebService[] webServices) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/WebServiceDefinition.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * A web service can call another web service to get some data. See {@link Request#localConnector()}
+ * provided by {@link RequestHandler#handle(Request, Response)}.
+ *
+ * @since 6.5
+ */
+@ServerSide
+@ExtensionPoint
+public interface WebServiceDefinition {
+  WebService.Controller define();
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/WebServiceDefinition.java]: MINOR Remove this use of "WebService"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 12 issues
* ![MINOR][MINOR] 9 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [WebServiceEngine.java#L94](https://github.com/SonarSource/sonarqube/blob/8fa63ba5deaf3e492bcaaeee79269f9fdf34580f/server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java#L94): Remove this use of "WebService"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [WebServiceEngine.java#L108](https://github.com/SonarSource/sonarqube/blob/8fa63ba5deaf3e492bcaaeee79269f9fdf34580f/server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java#L108): Remove this use of "WebService"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [WebServiceEngine.java#L138](https://github.com/SonarSource/sonarqube/blob/8fa63ba5deaf3e492bcaaeee79269f9fdf34580f/server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java#L138): Remove this use of "WebService"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [WebServiceEngine.java#L141](https://github.com/SonarSource/sonarqube/blob/8fa63ba5deaf3e492bcaaeee79269f9fdf34580f/server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java#L141): Remove this use of "WebService"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by simonbrandhof** [comment]
As discussed the API should not be deprecated. I'm closing this PR. Thanks.

**Pull request closed without merging**